src/chunk.c

```diff
@@ -837,6 +847,57 @@ chunk_find(Hyperspace *hs, Point *p)
 	return chunk;
 }
 
+static bool
+append_chunk_oid(ChunkScanCtx *scanctx, Chunk *chunk)
+{
+	if (chunk_is_complete(scanctx, chunk))
+	{
+		/* Fill in the rest of the chunk's data from the chunk table */
+		chunk_fill_stub(chunk, false);
+		scanctx->data = lappend_oid(scanctx->data, chunk->table_id);
+		return true;
+	}
+	return false;
+}
+
+List *
+chunk_find_all_oids(Hyperspace *hs, List *dimension_vecs, LOCKMODE lockmode)
+{
+	List	   *oid_list = NIL;
+	ChunkScanCtx ctx;
+	ListCell   *lc;
+
+	/* The scan context will keep the state accumulated during the scan */
+	chunk_scan_ctx_init(&ctx, hs, NULL);
+
+	/* Do not abort the scan when one chunk is found */
+	ctx.early_abort = false;
+
+	/* Scan all dimensions for slices enclosing the point */
+	foreach(lc, dimension_vecs)
+	{
+		DimensionVec *vec = lfirst(lc);
+
+		dimension_slice_and_chunk_constraint_join(&ctx, vec);
+	}
+
+	ctx.data = NIL;
+	chunk_scan_ctx_foreach_chunk(&ctx, append_chunk_oid, 0);
+	oid_list = ctx.data;
+
+	chunk_scan_ctx_destroy(&ctx);
+
+	foreach(lc, oid_list)
```

Comment:
Why not also do this work (locking) in `append_chunk_oid` (which is what I meant in previous comment)? You are still iterating twice here and then I presume once more when creating the appendInfos. That's at least three iterations of the same data. Ideally, you'd do all work in one iteration. Any reason not to?

Reply:
fixed